### PR TITLE
crush: weight_set and id remapping

### PR DESCRIPTION
**Pull request by dachary**

### CODE REVIEW

**Comment by dachary** [comment]
@liewegas here is a proposal for a minimalist integration of the weight_set in Ceph.

For buckets where the weight_set is not empty, the functions CrushWrapper::{insert_item,move_bucket,update_item,remove_item,remove_item_under} set the weight set and the target weight of new items to zero, and they do not adjust the weights of their parents. This allows the user to maintain the weights via the crushmap using the "choose_args" keyword::

```
// weight_set for pool 0
choose_args 0 {
  {
    bucket_id -5
    weight_set {
       position 0 weights 0x10000 0x20000
       position 1 weights 0x30000 0x20000
    }
    ids -20 -30 -25
  }

  {
    bucket_id -20
    weight_set {
       position 0 weights 0x10000 0x20000
       position 1 weights 0x30000 0x20000
    }
    ids -45 -23
  }
}

```
Whenever a choose_args is added to CrushWrapper (by compiling or decoding the crushmap), the buckets for which there exists at least one non empty weight_set are marked as such so that CrushWrapper::{insert_item,move_bucket,update_item,remove_item,remove_item_under} know to treat them differently.

The CrushWrapper::do_rule function takes an additional argument which is used as an index to find the
corresponding choose_args and OSDMap::_pg_to_raw_osds uses the pool number for this argument so that there can be a different choose_args per pool.

I'm moderately happy about this proposal but that's what I have so far ;-)

**Comment by liewegas** [comment]
I like this.  It's simple.  Basically weight changes just don't have any 
effect until whatever created the weight_set in the first place goes and 
runs again.

Where do you want to store the choose_args/weight_set?  As part of the map 
proper at the end of CrushWrapper::{encode,decode}?  Or as a separate type 
that the OSDMap manages/stores separately?

Will it be decompiled/compiled as part of the crush map (with a syntax 
like that below)?

**Comment by dachary** [comment]
> Where do you want to store the choose_args/weight_set? As part of the map proper at the end of CrushWrapper::{encode,decode}?

yes

> Will it be decompiled/compiled as part of the crush map (with a syntax like that below)?

yes

What bothers me is that I can't quite imagine how it will evolve beyond this minimalist proposal. But it's probably a problem for later :-)

**Comment by liewegas** [comment]
Hmm, what happens if you change the structure of the tree?  E.g., move a 
host to a different rack?  The weight_set isn't expressive enough to 
reflect a different size of the bucket, so it has to remove it form 
the old location, but it'll be 0'ed out in the new location.

**Comment by dachary** [comment]
when adding/removing items in a bucket, each choose_args[-1-bucket_id] must also be modified to remove the corresponding item or add an item with weight zero. This ensure the items count for each choose_args[-1-bucket_id] matches the items count of the map->bucket[-1-bucket_id] in the crushmap. In other words they must be kept in sync at all times. This should not be too much of a problem since adding / removing items/buckets is done via the CrushWrapper API. Unless there is something I'm not seeing ?

**Comment by liewegas** [comment]
yeah, that makes sense.  it just means that if you move a bucket between 2 
buckets (or an osd to another bucket) that it will be effectively removed 
from the old location but not added to the new one (until someone 
recalculates the weight_set values).

**Comment by dachary** [comment]
> it will be effectively removed from the old location but not added to the new one (until someone recalculates the weight_set values).

Yes. From a practical point of view this is not a useful behavior. Maybe it would be better to just forbid adding/removing items/buckets via the CrushWrapper API when there are weight_sets. The user has to add/remove items/buckets by decompiling/modifying/compiling the crushmap and is responsible for maintaining the weight_set accordingly.

**Comment by dachary** [comment]
The compiler part is implemented and the syntax is slightly simpler than what was originally proposed https://github.com/ceph/ceph/pull/14486/commits/843855616784fd56e8be559cee26afdcfb3cac73#diff-6c59ef5804a9ab71015e059395d84feeR107 . Please let me know if you see something odd ;-)

**Comment by dachary** [comment]
@liewegas updated the implementation with compile/decompile/encode/decode and a crushtool based test, moving on to disabling crush map updates via the CrushWrapper api when weight sets are present.

**Comment by dachary** [comment]
@liewegas I think it is ready for review :-)

**Comment by dachary** [comment]
<pre>teuthology-suite -k distro --priority 1000 --suite rados --subset $(expr $RANDOM % 200)/200 --email loic@dachary.org --ceph wip-multiweight --machine-type smithi</pre>

* **running** http://pulpito.ceph.com/loic-2017-04-16_20:21:30-rados-wip-multiweight-distro-basic-smithi/

File: src/crush/CrushWrapper.cc [hunk 1]
```diff
@@ -1540,6 +1572,38 @@ void CrushWrapper::decode(bufferlist::iterator& blp)
 	class_rname[c.second] = c.first;
       ::decode(class_bucket, blp);
       cleanup_classes();
+
```
Comment (liewegas) [inline src/crush/CrushWrapper.cc]: can you add a } if (!blp.end()) { here?  i have the class code deployed on a test cluster.

File: src/test/cli/crushtool/choose-args.crush [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+      [ 3.000 ]
+    ]
+  }
+  {
+    bucket_id -3
+    weight_set [
+      [ 1.000 2.000 5.000 ]
+      [ 3.000 2.000 5.000 ]
+    ]
+    ids -20 -30 -25
```
Comment (liewegas) [inline src/test/cli/crushtool/choose-args.crush]: weight_set uses [ ] for the lists.. maybe ids should too?

ids [ -20 -30 -25 ]

**Comment by dachary** [comment]
@liewegas the ids syntax was changed [here](https://github.com/ceph/ceph/pull/14486/commits/dbe36e08be00c6519a8c89718dd47b0219c20516#diff-407ca5f9de60dee8760a9d53ae9c6442R175) and tested [here](https://github.com/ceph/ceph/pull/14486/commits/dbe36e08be00c6519a8c89718dd47b0219c20516#diff-0057f181edd3554c94feabcb1586cbd7R116)

The extra [blp.end is here](https://github.com/ceph/ceph/pull/14486/commits/dbe36e08be00c6519a8c89718dd47b0219c20516#diff-c1ea4e8ae4b5520080ef65c1bd8ed666R1576)

Also added 57d4c8d and  18245ec fixing test leakage & valgrind complaints.

**Comment by liewegas** [comment]
Looks good!

Maybe we should deprecate the tree buckets more explicitly somehow so they can be removed.. maybe a bool CrushWrapper::has_tree_buckets(), and a health warning in luminous if they're in use to warn that support will be removed post-luminous?

**Comment by dachary** [comment]
good idea created http://tracker.ceph.com/issues/19654 to not forget about it.

**Comment by dachary** [comment]
<pre>teuthology-suite -k distro --priority 1000 --suite rados --subset $(expr $RANDOM % 200)/200 --email loic@dachary.org --ceph wip-multiweight --machine-type smithi</pre>

* **running** http://pulpito.ceph.com/loic-2017-04-18_13:14:34-rados-wip-multiweight-distro-basic-smithi/

**Pull request merged**